agent/src/config/config.h

```diff
@@ -212,73 +49,96 @@ LOG_DECLARE_CLASS_CONTEXT("CFG", "SoftwareContainer general config");
-     * @brief getBooleanValue Get a config value of type boolean
-     *
-     * @param group The name of the config group the value belongs to
-     * @param key The key for the config value
-     *
-     * @return A bool with the config value
-     */
-    bool getBooleanValue(const std::string &group, const std::string &key) const;
-
-    // Illegal values used as initial command line option values to check if user
-    // set them or not
-    static const std::string SC_CONFIG_PATH_INITIAL_VALUE;
-    static const int PRELOAD_COUNT_INITIAL_VALUE;
-    static const bool KEEP_CONTAINERS_ALIVE_INITIAL_VALUE;
-    static const int SHUTDOWN_TIMEOUT_INITIAL_VALUE;
-    static const std::string SERVICE_MANIFEST_DIR_INITIAL_VALUE;
-    static const std::string DEFAULT_SERVICE_MANIFEST_DIR_INITIAL_VALUE;
-    static const bool USE_SESSION_BUS_INITIAL_VALUE;
-
-    // Config group "SoftwareContainer"
-    static const std::string SC_GROUP;
-
-    // Config keys for SoftwareContainer group
-    static const std::string PRELOAD_COUNT_KEY;
-    static const std::string KEEP_CONTAINERS_ALIVE_KEY;
-    static const std::string USE_SESSION_BUS_KEY;
-    static const std::string SHUTDOWN_TIMEOUT_KEY;
-    static const std::string SHARED_MOUNTS_DIR_KEY;
-    static const std::string LXC_CONFIG_PATH_KEY;
-    static const std::string SERVICE_MANIFEST_DIR_KEY;
-    static const std::string DEFAULT_SERVICE_MANIFEST_DIR_KEY;
+    int getIntValue(const std::string &group, const std::string &key) const;
+    bool getBoolValue(const std::string &group, const std::string &key) const;
 
 private:
-    template<typename T>
-    T getValue(const std::string &group,
-               const std::string &key,
-               const std::map<std::string, T> &options) const;
+    void checkDependencies();
+    void checkForMandatory();
+    void getConfigsFromSources();
+    StringConfig prioritizedStringConfig(const std::vector<StringConfig> &configs) const;
+    IntConfig prioritizedIntConfig(const std::vector<IntConfig> &configs) const;
+    BoolConfig prioritizedBoolConfig(const std::vector<BoolConfig> &configs) const;
+
+    std::vector<std::unique_ptr<ConfigSource>> m_sources;
+    MandatoryConfigs m_mandatory;
+    ConfigDependencies m_dependencies;
+    std::vector<StringConfig> m_stringConfigs;
+    std::vector<IntConfig> m_intConfigs;
+    std::vector<BoolConfig> m_boolConfigs;
+};
 
-    template<typename T>
-    T getGlibValue(const std::string &group,
-                   const std::string &key) const;
 
-    /**
-     * @brief Load main config file source into member m_config
-     *
-     * @throws ConfigError On error
-     */
-    void loadConfig();
 
-    std::unique_ptr<ConfigLoaderAbstractInterface> m_loader;
-    std::unique_ptr<Glib::KeyFile> m_config;
-    std::unique_ptr<ConfigDefaults> m_defaults;
-    std::map<std::string, std::string> m_stringOptions;
-    std::map<std::string, int> m_intOptions;
-    std::map<std::string, bool> m_boolOptions;
-};
+
```

Comment:
I'm going to work this into the proper comments. It's not fully correct but it can be partly reused so I just kept it here for now, will be gone before merge.